### DISCUSSION TITLE
eslint not working (linter-eslint-node)

### DISCUSSION THREAD

**Discussion by bunyevacz** [pulsar-edit/.github#257] [Community Packages]
Hi,

I would like to use eslint for my javascript project to ease development. I tried every eslint package, but only `linter-eslint-node` worked, although briefly.

Now at every pulsar start I get the following popup error:
```
linter-eslint-node needs to install dependencies

linter
Install dependency?
[Yes] [No thanks] [Never]
```
I don't want to install eslint globally, I much prefer to use it local to the project.

Software details:
```
About Pulsar page:
Kubuntu 23.10
pulsar 1.114.2024021608
Electron 12.2.3
Chrome: 89.0.4389.128
Node: v14.16.0
```
There is a linter debug output (`Linter Eslint Node: Debug`):
```
Atom version: 1.114.2024021608
linter-eslint-node version: 1.0.5
Worker using Node at path: /usr/bin/node
Worker Node version: v18.13.0
Worker PID: 2059584
ESLint version: 8.57.0
ESLint location: /home/boognevatz/Desktop/newprojects/bcard/node_modules/.pnpm/eslint@8.57.0/node_modules/eslint/
Linting in this project performed by: linter-eslint-node
Hours since last Atom restart: 0.2
Platform: linux
Current file's scopes: [
  "source.js",
  "meta.block.function.js",
  "support.other.function.js"
]
linter-eslint-node configuration: {
  "advanced": {
    "useCache": false,
    "disableEslintIgnore": false,
    "showRuleIdInMessage": true
  },
  "warnAboutOldEslint": false,
  "scopes": [
    "source.js",
    "source.jsx",
    "source.js.jsx",
    "source.flow",
    "source.babel",
    "source.js-semantic",
    "source.ts"
  ],
  "nodeBin": "node",
  "autofix": {
    "fixOnSave": false,
    "rulesToDisableWhileFixing": [],
    "ignoreFixableRulesWhileTyping": false
  },
  "disabling": {
    "disableWhenNoEslintConfig": true,
    "rulesToSilenceWhileTyping": []
  }
}
```
Here is how I created the project:
```
sudo apt install npm
sudo npm install -g pnpm
pnpm create vite
# project name: bcard
# Select framework: React
# Select variant: javascript
cd bcard
pnpm run lint
```

I made an obvious error in source file, to test if esling works:
```
$ pnpm run lint
> bcard@0.0.0 lint /home/boognevatz/Desktop/newprojects/tmp/bcard
> eslint . --ext js,jsx --report-unused-disable-directives --max-warnings 0


/home/boognevatz/Desktop/newprojects/tmp/bcard/src/App.jsx
  1:10  error  'useState' is defined but never used  no-unused-vars
  7:29  error  'usState' is not defined              no-undef

✖ 2 problems (2 errors, 0 warnings)

 ELIFECYCLE  Command failed with exit code 1.
```

So I started pulsar, it worked once (there where 3 clickable buttons in status bar: 2 errors, 0 warning), I could click in the statusbar and I get a nice linter panel.

But it is not working anymore. I get the annoying popup at each start, and it does not work.

Looking at the `linter-eslint-node` sourcecode it looks like there is a key role of `.eslintignore` file, so I created one in project directory, but nothing changed.

Couple of interesting things:
- linter-eslint-node is not updated 2 years, eslint on the other hand changed drastically (config files)
- linter-eslint-node looks for `.eslintrc` file, eslint now uses `.eslintrc.cjs`/`eslintrc.js`/`.eslintrc.json`. And with upcoming eslint version 9.0, the configuration filename will change yet again (`eslint.config.js`/`eslint.config.mjs`)

I'm a little bit stuck how to debug further. Any pointer is much appreciated.

**Comment by savetheclocktower** [comment]
Have you tried clicking “Yes” on that popup? If so, what happens? That's where it typically installs `linter`, `busy-signal`, and maybe a couple other packages. If you didn't want it to do that, you could install them yourself and the effect would be the same.

I wrote most of `linter-eslint-node`, but since I wrote it, the `AtomLinter` organization on GitHub has been discontinued. Another developer and I did a small amount of planning for its future, but I've been busy, and the momentum has been lost. I suppose we'll have to give it another shot. In the meantime, though, it should be working fine for version 8.

**Reply by bunyevacz** [reply]
is it not able to pick up the local eslint installation (what is inside the node_modules/ directory) ?

Would be even nicer to use the command line, which is specified inside the package.json file:
```> eslint . --ext js,jsx --report-unused-disable-directives --max-warnings 0```


The other question is, how it can display the correct eslint version in the debug command, yet fail to use it?

I will start a config pulsar fresh, and report back. Right now with:
- pulsar-ide-typescript-alpha (`[X] Include Javascript` in the config panel)
- pulsar-outline-view
- linter
- linter-ui-default

I have linter, although looks like it is for typescript, but it is good enough for javascript as is. It produce different output than the command line version, but catch the bulk of the errors. 
(the main thing, is to be able to click on the error, and jump the editor's cursor to the line in question. From the terminal, it is a manual lookup)

I really need to advance with my javascript project, so I need to have something workish.

**Reply by savetheclocktower** [reply]
> is it not able to pick up the local eslint installation (what is inside the node_modules/ directory) ?

Of course it is. What I'm trying to figure out here are the exact symptoms you're experiencing. Let me see if I understand:

* Despite having `linter` and `linter-ui-default` and `linter-eslint-node` installed, you are prompted to install `linter` every time you launch Pulsar
* Linting simply fails to happen (despite seeming to work when first installed) without any other symptoms

Do I have this right? If so, let's take them in turn:

* I'd suggest uninstalling and reinstalling `linter` to see if that makes the prompt go away.
* Judging from your debug output, I'm wondering if your version of node is truly at `/usr/bin/node`. Does that match what you get if you run `which node` from your project directory in your terminal?

    Also, I'm pretty sure you're not on the latest version of `linter-eslint-node`, so uninstall what you've got and run this from the terminal: `ppm install AtomLinter/linter-eslint-node`. The version on the Pulsar package registry isn't up to date. Then relaunch Pulsar and see if that makes a difference. If not, then you can go to the `linter-eslint-node` settings, expand the “Uncommon” section, and check the “Enable Logging” option; that'll log a bunch of stuff to the devtools console so that you can see what's going on.
